funsor/distributions.py

```diff
@@ -194,6 +215,75 @@ def _multinomial_infer_value_domain(cls, **kwargs):
 DirichletMultinomial._infer_value_domain = classmethod(_multinomial_infer_value_domain)
 
 
+###############################################
+# Converting PyTorch Distributions to funsors
+###############################################
+
+@to_funsor.register(torch.distributions.Distribution)
+def torchdistribution_to_funsor(pyro_dist, output=None, dim_to_name=None):
+    import funsor.distributions  # TODO find a better way to do this lookup
+    funsor_dist_class = getattr(funsor.distributions, type(pyro_dist).__name__.split("__")[-1])
+    params = [to_funsor(
+            getattr(pyro_dist, param_name),
+            output=funsor_dist_class._infer_param_domain(
```

Comment:
`Distribution._infer_param_domain` is used here to enable a generic `torchdistribution_to_funsor` conversion that works by converting `TorchDistribution` parameters to funsors and constructing `Distribution` funsors from the results.

Reply:
Do I understand correctly that this assumes all parameters are batched scalars? So e.g. `MultivariateNormal` and `LowRankMultivariateNormal` and `Dirichlet` will need special casing? If so do you think we could add an `assert .event_dim == 0`?

Reply:
`_infer_param_domain` looks at the `arg_constraint` for each parameter to infer their `output`s.  In the case of `MultivariateNormal`, `arg_constraints["loc"]` is a `constraints.RealVector` object, so `_infer_param_domain` knows that its output shape is `loc.shape[-1:]`.